enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,1051 @@
+
+- Etcd database shall be backed up fully, rather just creating a snapshot to preserve
+  history and other data that could be lost.
+- Kubeconfigs and certificates needs to be backup and restored in order to keep communication working.
+  MicroShift could regenerate them, but it would mean that kubeconfigs used to remotely access cluster
+  will become invalid.
+  - Following approach might mean that certificates might need to be amended to update the
+    Subject Alternative Names (SAN) list.
+
+Based on reasons above, it was decided that whole `/var/lib/microshift` will be backed up.
+
+### MicroShift version metadata persistence
+
+When MicroShift is up and running healthy, green script will persist version into a file 
+within data dir, e.g. `/var/lib/microshift/.version`:
+```plaintext
+4.14.0
+```
+
+Value will be used during data migration to decide if
+- migration can be skipped,
+- migration needs to be blocked,
+- migration can be attempted.
+
+Creating the file by green script will indicate what was the MicroShift version that
+successfully ran using the data.
+
+### Data migration ("upgrade" or "downgrade")
+
+Data migration is process of transforming data from one schema version to another.
+It includes following areas:
+- Storage migration - upgrade of Kubernetes objects (e.g. from `v1beta1` to `v1`)
+  - It's performed by reading Resource in older version and writing newer version
+  - MicroShift will reuse existing 
+    [Kube Storage Version Migrator](https://github.com/openshift/kubernetes-kube-storage-version-migrator)
+    and [its Operator](https://github.com/openshift/cluster-kube-storage-version-migrator-operator).
+- etcd schema - although it's not verify likely in near future
+  - etcd project documents how to migrate from v2 to v3, 
+    but we'll also ask OpenShift etcd team for guidance.
+
+When new ostree deployment is staged and booted, it might or might not feature different
+version of MicroShift. If it's different, it might be newer or older.
+To keep process of data migration, its maintenance, testing matrix sane, we'll only allow
+data migration in one direction: forward in regards to Y stream.
+For now, maximum allowed version skew is Y+1, but it might change in the future depending
+on upstream migration rules.
+
+It means that it will be possible to use different Z versions of MicroShift with the same
+data unless there's a breaking change making it impossible, in such case it should be documented.
+
+Given above we can define:
+- rollback as booting older deployment due to admin actions or unhealthy (red) boot
+- upgrade as MicroShift version change from X.Y to X.Y+1 and resulting in data migration
+- downgrade as MicroShift version change from X.Y to X.Y-1 (or older) which is unsupported
+
+Both rollback and downgrade maybe look similar in terms of version change,
+the difference is that for rollback a matching backups exists,
+whereas for downgrade a data migration would have to be performed
```

Comment:
Rollback is automated and involves restoring from a backup. Downgrade is manual and reuses the existing data, right? There's no data migration from newer to older versions. That's why we only support downgrade within the same Y stream of releases.

Reply:
Well, last discussions steered me to terms "rollback" (due to greenboot) and "manual rollback".
Downgrade would be staging and booting deployment (that features older Y stream) that previously didn't exist on the device, so there's no backup for it to restore from.

If we don't perform data migration for any Z version within the same Y stream, then can we call it a downgrade?

Maybe we need to explicitly redefine the terms, what you think about:
- automated rollback - greenboot triggered
- manual rollback - to any deployment existing on the device (backup must be present for the deployment), triggered by admin
- upgrade - booting deployment with newer microshift (regardless if it's Y or Z that changed)
- supported downgrade
  - booting NEW deployment with older microshift (X.Y must match, Z can be different)
  - booting EXISTING deployment that backup exists for (Y doesn't need to match)
- unsupported downgrade - booting NEW deployment with older microshift (Y is lower)

Reply:
I think the use case for downgrading for an existing deployment with a backup where Y does not match is the same as a manual rollback. We would restore from the backup and there would be data loss of anything that happened in the previous Y version before the rollback.

That leaves any case with a MicroShift version where x.y match but z does not. To make that work, the API and data formats for x.y need to be compatible with all z versions. That will be true of kube and etcd. It needs to be true for anything MicroShift generates, too.